tools/kata-agent-ctl/src/client.rs

```diff
@@ -0,0 +1,1193 @@
+    }
+}
+
+fn handle_builtin_cmd(
+    cfg: &Config,
+    options: &mut Options,
+    cmd: &str,
+    args: &str,
+) -> (Result<()>, bool) {
+    let f = match get_builtin_cmd_func(&cmd) {
+        Ok(fp) => fp,
+        Err(e) => return (Err(e), false),
+    };
+
+    f(cfg, options, &args)
+}
+
+// Execute the ttRPC specified by the first field of "line". Return a result
+// along with a bool which if set means the client should shutdown.
+fn handle_agent_cmd(
+    cfg: &Config,
+    client: &AgentServiceClient,
+    health: &HealthClient,
+    options: &mut Options,
+    cmd: &str,
+    args: &str,
+) -> (Result<()>, bool) {
+    let f = match get_agent_cmd_func(&cmd) {
+        Ok(fp) => fp,
+        Err(e) => return (Err(e), false),
+    };
+
+    let result = f(cfg, client, health, options, &args);
+    if result.is_err() {
+        return (result, false);
+    }
+
+    let shutdown = cmd == SHUTDOWN_CMD;
+
+    (Ok(()), shutdown)
+}
+
+fn interactive_client_loop(
+    cfg: &Config,
+    options: &mut Options,
+    client: &AgentServiceClient,
+    health: &HealthClient,
+) -> Result<()> {
+    let result = builtin_cmd_list(cfg, options, "");
+    if result.0.is_err() {
+        return result.0;
+    }
+
+    let mut repeat_count: i64 = 1;
+
+    loop {
+        let cmdline = match readline("Enter command") {
+            Ok(c) => c,
+            Err(e) => return Err(format!("failed to read line: {}", e)),
+        };
+
+        if cmdline == "" {
+            continue;
+        }
+
+        if cmdline.starts_with(CMD_REPEAT) {
+            repeat_count = get_repeat_count(&cmdline);
+            continue;
+        }
+
+        let (result, shutdown) = handle_cmd(cfg, client, health, repeat_count, options, &cmdline);
+        if result.is_err() {
+            return result;
+        }
+
+        if shutdown {
+            break;
+        }
+
+        // Reset
+        repeat_count = 1;
+    }
+
+    Ok(())
+}
+
+fn readline(prompt: &str) -> std::result::Result<String, String> {
+    print!("{}: ", prompt);
+
+    match io::stdout().flush() {
+        Ok(_) => (),
+        _ => return Err(format!("failed to flush")),
+    };
+
+    let mut line = String::new();
+
+    match std::io::stdin().read_line(&mut line) {
+        // Remove NL
+        Ok(_) => Ok(line.trim_end().to_string()),
+
+        _ => Err(format!("failed to read line")),
+    }
+}
+
+fn agent_cmd_health_check(
+    cfg: &Config,
+    _client: &AgentServiceClient,
+    health: &HealthClient,
+    _options: &mut Options,
+    _args: &str,
+) -> Result<()> {
+    let mut req = CheckRequest::default();
+
+    // value unused
+    req.set_service("".to_string());
+
+    let reply = match health.check(&req, cfg.timeout_nano) {
+        Ok(reply) => reply,
+        Err(e) => return Err(format!("{}: {:?}", ERR_API_FAILED, e)),
+    };
+
+    info!(sl!(), "response received";
+        "response" => format!("{:?}", reply));
+
+    Ok(())
+}
+
+fn agent_cmd_health_version(
```

Comment:
One more idiomatic way of doing what you're doing with these callbacks is to create a Trait that defines one function prototype. Then you define one structure per command, each structure implementing the Trait.

Reply:
True, but the trait would also need things like functions for the name, etc, so this seemed like the easiest way to iterate of the fixed set of commands.